src/plugins/expressions/common/execution/execution.ts

```diff
@@ -202,147 +205,143 @@ export class Execution<
-        args = resolvedArgs;
-        timeStart = this.execution.params.debug ? now() : 0;
-        const output = await this.race(this.invokeFunction(fn, input, resolvedArgs));
-
-        if (this.execution.params.debug) {
-          const timeEnd: number = now();
-          (link as ExpressionAstFunction).debug = {
-            success: true,
-            fn: fn.name,
-            input,
-            args: resolvedArgs,
-            output,
-            duration: timeEnd - timeStart,
-          };
-        }
+  invokeChain(chainArr: ExpressionAstFunction[], input: unknown): Observable<any> {
+    return of(input).pipe(
+      ...(chainArr.map((link) =>
+        switchMap((currentInput) => {
+          const { function: fnName, arguments: fnArgs } = link;
+          const fn = getByAlias(this.state.get().functions, fnName);
+
+          if (!fn) {
+            throw createError({
+              name: 'fn not found',
+              message: i18n.translate('expressions.execution.functionNotFound', {
+                defaultMessage: `Function {fnName} could not be found.`,
+                values: {
+                  fnName,
+                },
+              }),
+            });
+          }
 
-        if (getType(output) === 'error') return output;
-        input = output;
-      } catch (rawError) {
-        const timeEnd: number = this.execution.params.debug ? now() : 0;
-        const error = createError(rawError) as ExpressionValueError;
-        error.error.message = `[${fnName}] > ${error.error.message}`;
-
-        if (this.execution.params.debug) {
-          (link as ExpressionAstFunction).debug = {
-            success: false,
-            fn: fn.name,
-            input,
-            args,
-            error,
-            rawError,
-            duration: timeStart ? timeEnd - timeStart : undefined,
-          };
-        }
+          if (fn.disabled) {
+            throw createError({
+              name: 'fn is disabled',
+              message: i18n.translate('expressions.execution.functionDisabled', {
+                defaultMessage: `Function {fnName} is disabled.`,
+                values: {
+                  fnName,
+                },
+              }),
+            });
+          }
 
-        return error;
-      }
-    }
+          if (this.execution.params.debug) {
+            link.debug = {
+              args: {},
+              duration: 0,
+              fn: fn.name,
+              input: currentInput,
+              success: true,
+            };
+          }
 
-    return input;
+          const timeStart = this.execution.params.debug ? now() : 0;
+
+          // `resolveArgs` returns an object because the arguments themselves might
+          // actually have `then` or `subscribe` methods which would be treated as a `Promise`
+          // or an `Observable` accordingly.
+          return this.race(this.resolveArgs(fn, currentInput, fnArgs)).pipe(
```

Comment:
I think the steps here might deserve some explanation.
What do you think?

Reply:
The code itself seems clear. I haven't changed anything significantly. Do you think we should provide an explanation even though it wasn't here before?